### PR TITLE
Add initial support for neuron devices

### PR DESCRIPTION
**Pull request by bfontain**
# What does this PR do?

Add some initial support for neuron devices to composer using the pytorch xla backend.

This PR fixes a few issues:
- Renames is_tpu_installed to is_xla_installed and changes it to store state in a global (vs importing and deleting torch_xla, which could delete a needed import).
- Adds a new device type 'neuron' for training on neuron instances
- Fixes support for AMP on XLA.
- Removes use of xm.optimizer_step (which introduces a second gradient reduce given that the model gets auto wrapped in DDP/FSDP). xm.optimizer_step does 3 things: 1) gradient reduction 2) applying the optimizer 3) xm.mark_step. We already have xm.mark_step in the microbatching loop, so remove the one here improves performance slightly.

Tested using a small model on 2 neuron cores via torchrun. support for composer CLI to come.

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/callbacks/speed_monitor.py [hunk 1]
```diff
@@ -83,30 +86,47 @@
         'int8': 130e12,
         'int4': 260e12,
     },
+    'trn1': {
```
Comment (j316chuck) [inline composer/callbacks/speed_monitor.py:89]: Can we add a source sheet from where you got these numbers?

File: composer/callbacks/speed_monitor.py [hunk 2]
```diff
@@ -83,30 +86,47 @@
...
+        'fp32': 190e12 / 2,
+        'tf32': 190e12 / 2,
+        'fp16': 190e12 / 2,
+        'amp_fp16': 190e12 / 2,
```
Comment (j316chuck) [inline composer/callbacks/speed_monitor.py:93]: Curious if you know why fp16 == fp32?

File: composer/devices/device_neuron.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    When running on Trn, we automatically set `export PJRT_DEVICE=NEURON`.
+    More details.
+    """
+
+    name = 'neuron'
+    dist_backend = 'xla'
+
+    def __init__(self):
+        import torch_xla.core.xla_model as xm
+        os.environ["NEURON_CC_FLAGS"] = "--auto-cast=none"
```
Comment (j316chuck) [inline composer/devices/device_neuron.py:33]: Can we please add a comment on where this flag comes from?

File: composer/trainer/trainer.py [hunk 4]
```diff
@@ -1320,7 +1320,7 @@ def __init__(
         if self._train_data_spec is not None:
             self.state.set_dataloader(self._train_data_spec.dataloader, train_dataloader_label,
                                       train_subset_num_batches)
-            if isinstance(self.state.device, DeviceTPU):
+            if self.state.device.dist_backend == 'xla':
```
Comment (j316chuck) [inline composer/trainer/trainer.py:1323]:
```suggestion
            if isinstance(self.state.device, DeviceNeuron):
```

**Review by j316chuck — COMMENTED** [review]
Left a few comments. Thanks for this! Overall LGTM

File: composer/callbacks/speed_monitor.py (hunk 2 above)
Comment (bfontain) [reply composer/callbacks/speed_monitor.py:95]: Thanks for the catch, I forgot to commit a change, fp32 is actually 4 times slower than fp16.

File: composer/callbacks/speed_monitor.py (hunk 1 above)
Comment (bfontain) [reply composer/callbacks/speed_monitor.py:91]: Done.

File: composer/devices/device_neuron.py (hunk 3 above)
Comment (bfontain) [reply composer/devices/device_neuron.py]: Done.

File: composer/trainer/trainer.py (hunk 4 above)
Comment (bfontain) [reply composer/trainer/trainer.py:1323]: I'll like to keep this as self.state.device.dist_backend so it works for all xla based devices (rather than having to add more devices to the isinstance check if another one comes along), what do you think?

Comment (j316chuck) [reply composer/trainer/trainer.py:1323]: Ah got it, makes sense! Thanks

**Review by j316chuck — APPROVED** [review]
Thanks for adding this device @bfontain
